### ISSUE TITLE
3 node cluster running on ECS using awsvpc networking

### ISSUE THREAD

**Issue by juliengrondin** [hashicorp/consul#6802]
Is there a way for the nodes to discover eachother within ECS using awsvpc networking?
Currently I get errors:
[WARN] agent: Join LAN failed: No servers to join, retrying in 30s
[ERR] agent: failed to sync remote state: No cluster leader
[ERR] agent: Cannot discover LAN provider=aws tag_key=consul tag_value=member: discover-aws: GetInstanceIdentityDocument failed: EC2MetadataRequestError: failed to get EC2 instance identity document
It seems the nodes are unable to discover each other to join a cluster.

**Comment by crhino** [comment]
This should work, although I have not personally tried it on AWS. What values are you passing to the `retry-join` flag/configuration value? That would help us in understanding the full context of this.

**Comment by juliengrondin** [comment]
`-retry-join=provider=aws tag_key=consul tag_value=member`
Resources are tagged with key : value `consul : member`

Are there any plans to use the ECS api as opposed to relying on EC2 describe-instance?

**Comment by juliengrondin** [comment]
This is still an issue - are there any work arounds?

**Comment by alkalinecoffee** [comment]
Might help to add a little more background on this:

```
2020-04-15T16:12:09.326Z [INFO] agent: discover-aws: Region not provided. Looking up region in metadata...: cluster=LAN
2020-04-15T16:12:09.872Z [ERROR] agent: Cannot discover address: cluster=LAN address="provider=aws tag_key=consul-datacenter tag_value=us-east-1-staging" error="discover-aws: GetInstanceIdentityDocument failed: EC2MetadataRequestError: failed to get EC2 instance identity document
caused by: RequestError: send request failed
caused by: Get http://169.254.169.254/latest/dynamic/instance-identity/document: dial tcp 169.254.169.254:80: connect: invalid argument"  
2020-04-15T16:12:09.872Z [WARN] agent: Join cluster failed, will retry: cluster=LAN retry_interval=30s error="No servers to join"
```

In ECS, the metadata IP depends on which platform version we're running.  According [to the docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html), the `instance-identity` path is not available (it's only available in EC2).

It seems that we either need to add some logic to determine whether we are running in ECS and disregard this call, or perhaps add a flag to disable it.

As it stands now, I don't believe we can use autodiscovery in ECS using tags in this fashion.

**Comment by cocakohler** [comment]
Any updates on that issue? @juliengrondin did you find a workaround for ECS. I want to run consul on ECS Fargate and i'm having the same problem.

**Comment by devarshishah3** [comment]
@cocakohler, there is a [learn guide](https://learn.hashicorp.com/tutorials/cloud/consul-client-ecs-ec2) to help get started with Consul on ECS/EC2 for HCP. The workflow for deploying Consul clients would be the same for self-managed. Are you looking to deploy Consul servers on ECS/Fargate or is it just clients? Fargate does not support awsvpc type and daemon scheduling. One way to deploy Consul clients would be to have a Consul agent run as a container within the task definition. Would love to understand your requirements and desired workflow for ECS/Fargate

**Comment by iandelahorne** [comment]
x-posting from hashicorp/go-discover#61:

We're seeing this too on the 1.10.2 container image on Fargate. Our consul servers are on EC2 and discovery works great for clients on EC2 using `-retry-join "provider=aws tag_key=role tag_value=consul-server"`

However, if we try to deploy a client sidecar on Fargate, it does not work. Here's a snippet of the task definition:

```
  "containerDefinitions": [
    {
      "name": "consul",
      "image": "public.ecr.aws/hashicorp/consul:1.10.2",
      "essential": true,
      "entryPoint": ["/bin/sh", "-ec"],
      "command": [
        "ECS_IPV4=$(curl -s $ECS_CONTAINER_METADATA_URI_V4 | jq -r '.Networks[0].IPv4Addresses[0]')\n exec consul agent -advertise \"$ECS_IPV4\" -datacenter development -retry-join \"provider=aws tag_key=role tag_value=consul-server\" -data-dir /consul/data"
      ],
      "logConfiguration": {
        "logDriver": "awslogs",
        "options": {
          "awslogs-create-group": "true",
          "awslogs-group": "/ecs/consultest",
          "awslogs-region": "us-west-2",
          "awslogs-stream-prefix": "consul"
        }
      },
      "portMappings": [
        {
          "containerPort": 8300,
          "hostPort": 8300,
          "protocol": "tcp"
        },
        {
          "containerPort": 8300,
          "hostPort": 8300,
          "protocol": "udp"
        }
      ]
    },
  "placementConstraints": [],
  "requiresCompatibilities": [
    "FARGATE"
  ],
```

In the logs, we see:
```
[ERROR] agent: Cannot discover address: cluster=LAN address="provider=aws tag_key=role tag_value=development" error="discover-aws: GetInstanceIdentityDocument failed: EC2MetadataRequestError: failed to get EC2 instance identity document
caused by: RequestError: send request failed
caused by: Get "http://169.254.169.254/latest/dynamic/instance-identity/document": dial tcp 169.254.169.254:80: connect: invalid argument"
```

According to the [ECS task IAM role docs](https://docs.aws.amazon.com/AmazonECS/latest/userguide/task-iam-roles.html), inside ECS the container IAM role should be fetched from `http://169.254.170.2$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`

**Comment by aeshaynes** [comment]
The only work around I could come up with for now is to register each Fargate container with AWS Service Discovery, then you can call them with the --retry-join.

**Comment by Amier3** [comment]
Hey @aeshaynes 

It seems like this issue fell through the cracks, and we apologize for that. I'll revive this issue internally and reach out to some of our engineers that work on consul-ecs integration.  We'll make an effort to get this resolved ( and documented ) soon 👍

**Comment by fdr2** [comment]
I am able to achieve cloud auto discover without using cloud auto discover. As pointed out above, the cloud auto discover uses the EC2 endpoints for collecting service ips with specific tags, but this wont work in Fargate due to the data needing to be collected from a different api. Instead I have been doing this with consistent success in my docker entrypoint that simply invokes the consul docker-entrypoint at the end.
```
# Get the cluster service ARN
CLUSTER_ARN=$(curl --silent --show-error "${ECS_CONTAINER_METADATA_URI_V4}"|jq -r '.Labels["com.amazonaws.ecs.cluster"]')

SERVICE_ARNS=$(aws ecs list-services --cluster "$CLUSTER_ARN" | jq -r ".serviceArns[]" |grep -i consul)

TASK_ARNS=$(for I in $SERVICE_ARNS; do aws ecs list-tasks --cluster "$CLUSTER_ARN" --service-name "$I" | jq ".taskArns[]"; done | xargs) # no raw-jq so we preserve quotes for list

# shellcheck disable=SC2086 # TaskArn is required to be a list of strings
TASK_IPS=$(aws ecs describe-tasks --cluster "$CLUSTER_ARN" --tasks $TASK_ARNS | jq -r ".tasks[0:3] | .[] | select(.desiredStatus==\"RUNNING\").containers[] | select(.lastStatus==\"RUNNING\").networkInterfaces[].privateIpv4Address")

if [ -n "$TASK_IPS" ]; then
  # shellcheck disable=SC2034 # TaskArn is required to be a list of strings
  AUTO_JOIN=$(for I in $TASK_IPS; do echo -n "\"$I\", "; done)
fi
```

Keep in mind, there is a chance that a new container will enumerate N+1 ips, due to the replaced container still being present when the new container comes online, but this does not seem to stop the new service from finding the healthy nodes and restoring quorum.

I use the AUTO_JOIN variable above to inject the current cluster ips into my config on startup.
```
if [ -n "${AUTO_JOIN}" ]; then
  sed -i "s/^retry_join.*/retry_join = \[ ${AUTO_JOIN}\]/" /consul/config/consul.hcl
fi
```

Another thing I have noticed, is when using a Load Balancer (for admin management traffic) by default the `deployment_maximum_percent` is 200 and `deployment_minimum_healthy_percent` is 100. This will prevent a new Consul service from replacing a healthy Consul service because the new service will not pass the healthcheck (200 response from /v1/status/leader) until the old server has left the cluster. Instead, I set these values to 100 and 0, respectively in order to have Fargate shutdown the task to be replaced while creating the new one, thereby allowing the new service to pass healthchecks.

maybe not the best way, but hopefully food for thought for those of you still chewing on this.

**Comment by fdr2** [comment]
I submitted [PR #197 to go-discover](https://github.com/hashicorp/go-discover/pull/197) in order to provide ECS-Tag auto-discovery support

**Comment by Amier3** [comment]
Cross-posting a comment from the PR above:

> The team did want to caution though that we haven't evaluated the effects of running consul servers inside of ECS ( aside from a single node cluster for dev/testing ) . So for the moment we only officially support using [servers outside of ECS](https://www.consul.io/docs/ecs/requirements). There was a couple of factors that went into that decision -- one of them being some performance concerns around the EFS that ECS uses under the hood.

So we'll continue to try and resolve this issue to allow for consul ecs server experimentation

**Comment by pglass** [comment]
The ECS support is merged into go-discover. Next, Consul needs to be updated with the new version of the library.

(triggers auto-closed the issue on a merge, so reopened this)

**Comment by pglass** [comment]
The [support for ECS discovery](https://github.com/hashicorp/consul/pull/13782) in [Cloud-Auto Join strings](https://developer.hashicorp.com/consul/docs/install/cloud-auto-join#amazon-ec2-and-ecs) was included in the 1.12.9, 1.13.6, and 1.14.4 patch releases of Consul, so closing this issue.
